File: frontend/src/api/opportunities.ts
import { apiClient } from './client'

export type OpportunityStage = 'prospect' | 'qualify' | 'proposal' | 'negotiate' | 'closedWon'

export interface Opportunity {
  id: string
  accountId: string
  companyName: string
  regionCode: string
  tags: string
  amount: string
  ownerName: string
  stage: OpportunityStage
  isNoFollowUp: boolean
  winType?: 'first' | 'reorder' | null
  createdAt?: string
  updatedAt?: string
}

export interface OpportunityCreatePayload {
  accountId: string
  companyName: string
  regionCode: string
  tags: string
  amount: string
  ownerName: string
  stage: OpportunityStage
  isNoFollowUp?: boolean
  winType?: 'first' | 'reorder' | null
}

export interface OpportunityUpdatePayload {
  accountId?: string
  companyName?: string
  regionCode?: string
  tags?: string
  amount?: string
  ownerName?: string
  stage?: OpportunityStage
  isNoFollowUp?: boolean
  winType?: 'first' | 'reorder' | null
}

function fromApi(data: Record<string, unknown>): Opportunity {
  return {
    id: data.id as string,
    accountId: data.account_id as string,
    companyName: data.company_name as string,
    regionCode: data.region_code as string,
    tags: data.tags as string,
    amount: data.amount as string,
    ownerName: data.owner_name as string,
    stage: data.stage as OpportunityStage,
    isNoFollowUp: Boolean(data.is_no_follow_up),
    winType: (data.win_type as Opportunity['winType']) || null,
    createdAt: data.created_at as string | undefined,
    updatedAt: data.updated_at as string | undefined,
  }
}

function toApi(payload: OpportunityCreatePayload | OpportunityUpdatePayload): Record<string, unknown> {
  return {
    account_id: payload.accountId,
    company_name: payload.companyName,
    region_code: payload.regionCode,
    tags: payload.tags,
    amount: payload.amount,
    owner_name: payload.ownerName,
    stage: payload.stage,
    is_no_follow_up: payload.isNoFollowUp,
    win_type: payload.winType,
  }
}

export async function listOpportunities(): Promise<Opportunity[]> {
  const { data } = await apiClient.get('/opportunities')
  return (data as Record<string, unknown>[]).map(fromApi)
}

export async function getOpportunity(id: string): Promise<Opportunity> {
  const { data } = await apiClient.get(`/opportunities/${id}`)
  return fromApi(data as Record<string, unknown>)
}

export async function createOpportunity(payload: OpportunityCreatePayload): Promise<Opportunity> {
  const { data } = await apiClient.post('/opportunities', toApi(payload))
  return fromApi(data as Record<string, unknown>)
}

export async function updateOpportunity(id: string, payload: OpportunityUpdatePayload): Promise<Opportunity> {
  const { data } = await apiClient.patch(`/opportunities/${id}`, toApi(payload))
  return fromApi(data as Record<string, unknown>)
}

export async function deleteOpportunity(id: string): Promise<void> {
  await apiClient.delete(`/opportunities/${id}`)
}

File: backend/app/routers/opportunities.py
import uuid
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import or_
from sqlalchemy.orm import Session
from typing import List, Optional
from app.db import get_db
from app import models, schemas
from app.core.security import get_current_active_user

router = APIRouter(
    prefix="/opportunities",
    tags=["opportunities"],
    dependencies=[Depends(get_current_active_user)],
)


def _opportunity_scope(query, user: models.User):
    scope = user.role_obj.record_access_scope if user.role_obj else "all"
    if scope == "own":
        return query.join(models.Account, models.Opportunity.account_id == models.Account.id).filter(
            models.Account.owner_id == user.id
        )
    if scope in ("team", "market"):
        return query.join(models.Account, models.Opportunity.account_id == models.Account.id).filter(
            models.Account.market_code == user.market_code
        )
    return query


@router.get("", response_model=List[schemas.OpportunityOut])
def list_opportunities(
    stage: Optional[str] = Query(None),
    q: Optional[str] = Query(None),
    region_code: Optional[str] = Query(None),
    user: models.User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
):
    query = _opportunity_scope(db.query(models.Opportunity), user)
    if stage:
        query = query.filter(models.Opportunity.stage == stage)
    if region_code:
        query = query.filter(models.Opportunity.region_code.ilike(f"%{region_code}%"))
    if q:
        like = f"%{q}%"
        query = query.filter(
            or_(
                models.Opportunity.company_name.ilike(like),
                models.Opportunity.region_code.ilike(like),
                models.Opportunity.tags.ilike(like),
            )
        )
    return query.order_by(models.Opportunity.created_at.desc()).all()


@router.get("/{opportunity_id}", response_model=schemas.OpportunityOut)
def get_opportunity(opportunity_id: str, db: Session = Depends(get_db)):
    opp = db.query(models.Opportunity).filter(models.Opportunity.id == opportunity_id).first()
    if not opp:
        raise HTTPException(status_code=404, detail="Opportunity not found")
    return opp


@router.post("", response_model=schemas.OpportunityOut, status_code=201)
def create_opportunity(payload: schemas.OpportunityCreate, db: Session = Depends(get_db)):
    data = payload.model_dump(exclude_unset=True)
    if not data.get("id"):
        data["id"] = f"oc-{uuid.uuid4().hex[:8]}"
    opp = models.Opportunity(**data)
    db.add(opp)
    db.commit()
    db.refresh(opp)
    return opp


@router.patch("/{opportunity_id}", response_model=schemas.OpportunityOut)
def update_opportunity(opportunity_id: str, payload: schemas.OpportunityUpdate, db: Session = Depends(get_db)):
    opp = db.query(models.Opportunity).filter(models.Opportunity.id == opportunity_id).first()
    if not opp:
        raise HTTPException(status_code=404, detail="Opportunity not found")
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(opp, key, value)
    db.commit()
    db.refresh(opp)
    return opp


@router.delete("/{opportunity_id}", status_code=204)
def delete_opportunity(opportunity_id: str, db: Session = Depends(get_db)):
    opp = db.query(models.Opportunity).filter(models.Opportunity.id == opportunity_id).first()
    if not opp:
        raise HTTPException(status_code=404, detail="Opportunity not found")
    db.delete(opp)
    db.commit()
    return None
